lib/charms/glauth_k8s/v0/ldap.py

```diff
@@ -320,28 +322,38 @@ def _on_relation_broken(self, event: RelationBrokenEvent) -> None:
         )
         secret.remove()
 
+    def get_app_password(self, relation_id: int) -> Optional[str]:
+        """Retrieve the secret for a given relation."""
+        try:
+            secret = self.charm.model.get_secret(
+                label=BIND_ACCOUNT_SECRET_LABEL_TEMPLATE.substitute(relation_id=relation_id)
+            )
+        except SecretNotFoundError:
+            return ""
+        return secret.get_content().get("password")
+
     def update_relations_app_data(
-        self, /, data: Optional[LdapProviderBaseData] = None, relation_id: Optional[int] = None
+        self,
+        data: Union[LdapProviderBaseData, LdapProviderData],
+        /,
+        relation_id: Optional[int] = None,
     ) -> None:
         """An API for the provider charm to provide the LDAP related information."""
-        if data is None:
-            return
-
         if not (relations := self.charm.model.relations.get(self._relation_name)):
             return
 
-        if relation_id is not None:
+        if relation_id is not None and getattr(data, "bind_password", None):
```

Comment:
Does this (`getattr(data, "bind_password", None)`) check if the `data` is typed `LdapProviderData`?

If so, I guess probably using `isinstance(data, LdapProviderData)` is more robust.

Reply:
You are right, `bind_password` is mandatory so it can't be null. I will change this.

Thanks